esp32/s2/c3: power management: initial support

glaubermaroto opened the pull request:
This PR brings initial System Power Management support for all Espressif SoCs currently maintained on Zephyr (i.e., ESP32, ESP32-S2, and ESP32-C3).

Most of the low-power handling is under RTC components. Thus, partial RTC initialization and configuration are also brought here. More specifically, this PR brings:

- Initial System Power Management policy implementation;
- Deep sleep mode support;
- Light sleep mode support
- Timer, RTC_IO (EXT0/EXT1), and GPIO as wake-up sources;
- Sample codes for both sleep modes.

Aside from the above, the sample code showcases a way to extract the wake-up reason, how to set a wake-up source and how to trigger deep sleep.

Commit:
pm: esp32: system power management

- Adds initial system power management support.
- Adds option to add extra delay when waking from
deep sleep.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32: set RTC registers to known state

Low-power management is part of the RTC peripherals' domain
on ESP32. This dependency implies the need to bring some RTC
registers to a known state, during system initialization, to
achieve proper low-power handling.

The RTC slow memory region is also delimited and used during
power domain options selection.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
esp32: samples: power management: add deep sleep sample

Add deep sleep sample code with support for the
following wake-up sources:

- Timer
- EXT0/EXT1: IO under RTC domain
- GPIO (ESP32-C3 only)

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32s2: set RTC registers to known state

Low-power management is part of the RTC peripherals' domain
on ESP32S2. This dependency implies the need to bring some RTC
registers to a known state, during system initialization, to
achieve proper low-power handling.

The RTC slow memory region is also delimited and used during
power domain options selection.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
pm: esp32s2: system power management

Adds initial system power management support.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
pm: esp32c3: system power management

Adds initial system power management support.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32c3: set RTC registers to known state

Low-power management is part of the RTC peripherals' domain
on ESP32C3. This dependency implies the need to bring some RTC
registers to a known state, during system initialization, to
achieve proper low-power handling.

The RTC slow memory region is also delimited and used during
power domain options selection.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32: add light sleep sample code

Add light sleep sample code.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
west: hal_espressif: update revision

Update HAL manifest revision.

This brings power management hal_espressif support.
Fixes #52073

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
doc: services: pm: add reference to esp32 sample

Includes reference to esp32 deep sleep sample code
to system power management documentation,

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
soc: esp32x: replace STATUS by int

Replace 'STATUS' by 'int' in extern function
signatures.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

ceolin commented on soc/xtensa/esp32/power.c:
```diff
@@ -0,0 +1,58 @@
+	 * System is now in active mode. Reenable interrupts which were disabled
+	 * when OS started idling code.
+	 */
+	__asm__ volatile ("waiti 0");
```
is it really re-enabling interruptions ?

glaubermaroto replied:
Not really, I have removed it.

ceolin commented:
Is there a reason to this state not be defined in the device tree ? So the policy can automatically put the device to sleep when the system goes idle ? There are targets where these deep sleep states are supposed to be only triggered by the application (using the `force` api), just trying to understand if that is the case here.

Commit:
boards: xtensa: esp32: dts: add power states

Add power states to ESP32 board dts.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: xtensa: esp32s2_saola: dts: add power states

Add power states to ESP32-S2 Saola board dts.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

Commit:
boards: riscv: esp32c3_devkitm: dts: add power states

Add power states to ESP32-C3 Devkit-M board dts.

Signed-off-by: Glauber Maroto Ferreira <glauber.ferreira@espressif.com>

glaubermaroto commented:
> Is there a reason to this state not be defined in the device tree ? So the policy can automatically put the device to sleep when the system goes idle ? There are targets where these deep sleep states are supposed to be only triggered by the application (using the `force` api), just trying to understand if that is the case here.

No particular reason, thanks for pointing this out. I have updated the PR to include power states' definitions on the device tree.

glaubermaroto commented:
I have rebased to solve conflict and included power states definitions on the device tree.

glaubermaroto commented:
Rebased and fixed the failing CI test.

glaubermaroto commented:
@ceolin, could you take another look, please?

ceolin commented on soc/xtensa/esp32s2/power.c:
```diff
@@ -0,0 +1,46 @@
+		break;
+	}
+
+	k_cpu_idle();
```
why are you calling `k_cpu_idle()` here ? This will be called after the system wakes up and does not look correct, you should just return and let the idle thread runs again. The pm subsystem then will check with the policy manager which power state to use, if there is no suitable state it will return `PM_STATE_ACTIVE` and the idle thread will then call `k_cpu_idle()`.

glaubermaroto replied:
> I have just two concerns, one in `pm_state_set` that is calling `k_cpu_idle` and the other is about not `pm_state_exit_post_ops` not being restoring interruptions. Other than that looks great this pr.

I have removed the `k_cpu_idle` call and included interruption restoring for the PM_STANDBY power state. When SYSTEM_OFF is used, SoC (re)initialization already initializes and restores interruptions.

ceolin commented on soc/xtensa/esp32s2/power.c:
```diff
@@ -0,0 +1,46 @@
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
```
At this point you should restore interruptions, the system expects that interruptions are restored after this callback.

glaubermaroto replied:
> I have just two concerns, one in `pm_state_set` that is calling `k_cpu_idle` and the other is about not `pm_state_exit_post_ops` not being restoring interruptions. Other than that looks great this pr.

I guess I have addressed those 2 points now. I have also reinserted the forced state on the deep sleep sample code not to have to tweak individually to each SoC.

ceolin reviewed (commented):
I have just two concerns, one in `pm_state_set` that is calling `k_cpu_idle` and the other is about not `pm_state_exit_post_ops` not being restoring interruptions. Other than that looks great this pr.

glaubermaroto commented:
@ceolin I have addressed the points you mentioned and took the opportunity also to include PM_STANDY support.

The CI failure seems unrelated to this PR.

ceolin commented:
> @ceolin I have addressed the points you mentioned and took the opportunity also to include PM_STANDY support.
> 
> The CI failure seems unrelated to this PR.

Thanks !

gmarull commented on soc/xtensa/esp32/Kconfig.defconfig:
```diff
@@ -54,4 +57,19 @@ config GDBSTUB_BUF_SZ
 
 endif
 
+# Indicates if a console UART is used
+config ESP_CONSOLE_UART
+	bool
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_CONSOLE)) && \
+		$(dt_chosen_has_compat,$(DT_CHOSEN_Z_CONSOLE),$(DT_COMPAT_ESPRESSIF_ESP32_UART))
+
+# During system clock initialization this config is used to avoid data lost
+# during APB frequency switching. Due to as ESP32 ROM bug, UART2 is not
+# supported for console output, hence, this case is not handled here.
+config ESP_CONSOLE_UART_NUM
```
why not use `help` field here?

glaubermaroto replied:
This config supports HAL Espressif's power management logic to flush the console's FIFO queue if needed. Once it is not meant to be user-tweaked, I am not sure if `help` is really needed.

gmarull replied:
I think `help` is just the natural place where to explain what a Kconfig symbol does.

sylvioalves replied:
@gmarull, sure, I'll update this.

gmarull commented on soc/xtensa/esp32s2/Kconfig.defconfig:
```diff
@@ -46,4 +49,19 @@ config MP_NUM_CPUS
 config MINIMAL_LIBC_OPTIMIZE_STRING_FOR_SIZE
 	default n
 
+# Indicates if a console UART is used
+config ESP_CONSOLE_UART
+	bool
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_CONSOLE)) && \
+	$(dt_chosen_has_compat,$(DT_CHOSEN_Z_CONSOLE),$(DT_COMPAT_ESPRESSIF_ESP32_UART))
+
+# During system clock initialization this config is used to avoid data lost
+# during APB frequency switching. Due to as ESP32 ROM bug, UART2 is not
+# supported for console output, hence, this case is not handled here.
+config ESP_CONSOLE_UART_NUM
+	int
+	default -1 if !ESP_CONSOLE_UART
+	default 1 if !$(dt_nodelabel_enabled,"uart0")
+	default 0
```
same, also, why is `ESP_CONSOLE_UART` needed? seems to be an auxiliary option?

glaubermaroto replied:
Ditto.

sylvioalves replied:
@gmarull, we need to find a way to get the UARTx number properly. Espressif's uart calls usually depends on this value (0, 1, and 2). Do you have a suggestion on how to link/get this number properly? Last time I tried adding this info into UART's driver as `peripheral_id` instance, it was not approved. Any recommendations?

sylvioalves replied:
@gmarull Any opinion regarding this? This is the blocking comment right now.

sylvioalves replied:
UART number will be handled in another PR.

gmarull commented on soc/riscv/esp32c3/power.c:
```diff
@@ -0,0 +1,52 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
```
which Kernel APIs are being used here?

sylvioalves replied:
Fixed, no API was used.

gmarull commented on soc/riscv/esp32c3/power.c:
```diff
@@ -0,0 +1,52 @@
+		/* Nothing to do. */
+		break;
+	case PM_STATE_STANDBY:
+		arch_irq_unlock(MSTATUS_IEN);
```
why `arch` version? please, include necessary headers as well.

sylvioalves replied:
Fixed and updated to `irq_unlock(MSTATUS_IEN);`.

gmarull commented on samples/boards/esp32/deep_sleep/src/main.c:
```diff
@@ -0,0 +1,121 @@
+#endif
+#endif
+
+static const struct pm_state_info residency_info[] =
```
where is this used?

gmarull commented on samples/boards/esp32/light_sleep/src/main.c:
```diff
@@ -0,0 +1,96 @@
+static const struct gpio_dt_spec button =
+	GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios, {0});
+
+static const struct pm_state_info pwr_state_info[] =
+	PM_STATE_INFO_LIST_FROM_DT_CPU(DT_NODELABEL(cpu0));
```
this is retrieving all states info just to obtain a residency time. You can just obtain that manually.

gmarull reviewed (dismissed):
a few minor changes/questions

sylvioalves committed:
boards: esp32_net: ignore PM tests

esp32_net is part of AMP solution and it is not intended
to have PM support. Ignore this board from tests

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
@gmarull @LucasTambor PTAL.

gmarull reviewed (commented):
no objections

knthm commented on soc/riscv/esp32c3/power.c:
```diff
@@ -0,0 +1,52 @@
+		break;
+	case PM_STATE_STANDBY:
+		irq_unlock(MSTATUS_IEN);
+		__asm__ volatile("wfi");
+		esp_light_sleep_start();
```
@glaubermaroto
This is a very strange hack. Is there a reason you didn't put light sleep in a separate SUSPEND_TO_IDLE state, but enable it directly after waking from standby?

I'm trying to [refactor the interrupt enable handling across all architectures in the context of System PM wake-up](https://github.com/zephyrproject-rtos/zephyr/pull/60998#discussion_r1339312370), and this section is blocking that.

I'll gladly create a PR for moving this logic into its own PM_STATE, provided there isn't an explicit reason for this.
